From c8dc3cec2da7cc7083c5c750ddad041dca352c72 Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Fri, 1 Sep 2023 18:03:27 +0200
Subject: [PATCH 0632/1695] drivers: counter: stm32_lptimer: add support for
 stm32mp25

Add support for stm32mp25 in stm32_lptimer driver. Use new compatible to
handle this new HW variant, even if no major change is expected on the
counter driver. In stm32mp25, lptimer has new channels. It also supports
input capture and autonomous mode.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 61ae595f46aca70ef3b5307a1b2e512430d2a513 drivers: counter: stm32_lptimer: add support for stm32mp25]
Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: I0f4d62e0f6624df2cb7722f01af5c18d0ba20576
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/326258
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/counter/stm32_lptimer.c | 68 +++++++++++++++++++++-------
 1 file changed, 52 insertions(+), 16 deletions(-)

diff --git a/core/drivers/counter/stm32_lptimer.c b/core/drivers/counter/stm32_lptimer.c
index b8df21920..f612b3556 100644
--- a/core/drivers/counter/stm32_lptimer.c
+++ b/core/drivers/counter/stm32_lptimer.c
@@ -115,6 +115,10 @@
 
 #define TIMEOUT_US_100MS	U(100000)
 
+struct lptimer_compat_data {
+	bool ier_wr_disabled;		/* disable lptimer to set dier */
+};
+
 struct lptimer_device {
 	const void *fdt;
 	int node;
@@ -122,6 +126,7 @@ struct lptimer_device {
 	struct lptimer_driver_data *ddata;
 	struct counter_device *counter_dev;
 	struct itr_handler *itr;
+	const struct lptimer_compat_data *cdata;
 };
 
 struct stm32_lptimer_config {
@@ -164,11 +169,14 @@ static TEE_Result stm32_lpt_counter_set_alarm(struct counter_device *counter)
 	}
 	io_write32(base + _LPTIM_ICR, _LPTIM_CMPARR_OK);
 
-	/*
-	 * LPTIM_IER register must only be modified
-	 * when the LPTIM is disabled
-	 */
-	io_clrbits32(base + _LPTIM_CR, _LPTIM_CR_ENABLE);
+	if (lpt_dev->cdata && lpt_dev->cdata->ier_wr_disabled) {
+		/*
+		 * LPTIM_IER register must only be modified
+		 * when the LPTIM is disabled
+		 */
+		io_clrbits32(base + _LPTIM_CR, _LPTIM_CR_ENABLE);
+	}
+
 	/* enable, Compare match Interrupt */
 	io_setbits32(base + _LPTIM_IER, _LPTIM_IXX_CMPM);
 
@@ -195,11 +203,13 @@ static TEE_Result stm32_lpt_counter_cancel_alarm(struct counter_device *counter)
 
 	cr = io_read32(base + _LPTIM_CR);
 
-	/*
-	 * LPTIM_IER register must only be modified
-	 * when the LPTIM is disabled
-	 */
-	io_clrbits32(base + _LPTIM_CR, _LPTIM_CR_ENABLE);
+	if (lpt_dev->cdata && lpt_dev->cdata->ier_wr_disabled) {
+		/*
+		 * LPTIM_IER register must only be modified
+		 * when the LPTIM is disabled
+		 */
+		io_clrbits32(base + _LPTIM_CR, _LPTIM_CR_ENABLE);
+	}
 	io_clrbits32(base + _LPTIM_IER, _LPTIM_IXX_CMPM);
 
 	/*
@@ -479,7 +489,10 @@ static TEE_Result stm32_lptimer_set_driverdata(struct lptimer_device *lpt_dev)
 }
 
 #ifdef CFG_EMBED_DTB
-#define LPTIMER_COUNTER_COMPAT	"st,stm32-lptimer-counter"
+static const char * const stm32_lptimer_counter_compat[] = {
+	"st,stm32-lptimer-counter",
+	"st,stm32mp25-lptimer-counter",
+};
 
 struct dt_id_attr {
 	/* The effective size of the array is meaningless here */
@@ -494,6 +507,7 @@ static TEE_Result stm32_lptimer_parse_fdt(struct lptimer_device *lpt_dev)
 	const void *fdt = lpt_dev->fdt;
 	int node = lpt_dev->node;
 	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned int i = 0;
 	int len = 0;
 
 	assert(lpt_dev && lpt_dev->fdt);
@@ -520,10 +534,16 @@ static TEE_Result stm32_lptimer_parse_fdt(struct lptimer_device *lpt_dev)
 		return res;
 
 	node = fdt_subnode_offset(fdt, node, "counter");
-	if (node >= 0 &&
-	    !fdt_node_check_compatible(fdt, node, LPTIMER_COUNTER_COMPAT) &&
-	    fdt_get_status(fdt, node) != DT_STATUS_DISABLED)
-		pdata->mode = LPTIMER_MODE_COUNTER;
+	if (node >= 0 && fdt_get_status(fdt, node) != DT_STATUS_DISABLED) {
+		for (i = 0; i < ARRAY_SIZE(stm32_lptimer_counter_compat); i++) {
+			const char *compat = stm32_lptimer_counter_compat[i];
+
+			if (!fdt_node_check_compatible(fdt, node, compat)) {
+				pdata->mode = LPTIMER_MODE_COUNTER;
+				break;
+			}
+		}
+	}
 
 	return 0;
 }
@@ -632,6 +652,7 @@ static TEE_Result stm32_lptimer_probe(const void *fdt, int node,
 
 	lpt_dev->fdt = fdt;
 	lpt_dev->node = node;
+	lpt_dev->cdata = compat_data;
 	res = probe_lpt_device(lpt_dev);
 
 	if (res)
@@ -640,8 +661,23 @@ static TEE_Result stm32_lptimer_probe(const void *fdt, int node,
 	return res;
 }
 
+static const struct lptimer_compat_data lptimer_cdata = {
+	.ier_wr_disabled = true,
+};
+
+static const struct lptimer_compat_data lptimer_mp25_cdata = {
+	.ier_wr_disabled = false,
+};
+
 static const struct dt_device_match stm32_lptimer_match_table[] = {
-	{ .compatible = "st,stm32-lptimer" },
+	{
+		.compatible = "st,stm32-lptimer",
+		.compat_data = &lptimer_cdata,
+	},
+	{
+		.compatible = "st,stm32mp25-lptimer",
+		.compat_data = &lptimer_mp25_cdata,
+	},
 	{ }
 };
 
-- 
2.47.2

